contracts/BSCValidatorSet.sol

```diff
@@ -344,6 +349,36 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
     return CODE_OK;
   }
 
+  function swap(address[] memory set, uint i, uint j) internal pure {
+    address tmp = set[i];
+    set[i]=set[j];
+    set[j]=tmp;
+  }
+
+  function shuffle(address[] memory validators, uint256 epochNumber, uint start, uint limit, uint modNumber) internal pure {
+    for (uint i = 0; i<limit; i++) {
+      uint random = uint(keccak256(abi.encodePacked(epochNumber, validators[i]))) % modNumber;
+      if ( i!=random ) {
+        swap(validators, start+i, start+random);
+      }
+    }
+  }
+
+  function getMiningValidators(uint256 epochNumber) public view returns(address[] memory) {
+    address[] memory validators = getValidators();
+    if (validators.length <= numOfCabinets) {
+      return validators;
+    }
+    
+    shuffle(validators, epochNumber, 0, maxNumOfWorkingCandidates, numOfCabinets);
+    shuffle(validators, epochNumber, numOfCabinets-maxNumOfWorkingCandidates, maxNumOfWorkingCandidates, validators.length-numOfCabinets+maxNumOfWorkingCandidates);
+    address[] memory miningValidators = new address[](numOfCabinets);
+    for (uint i=0;i<numOfCabinets;i++) {
+      miningValidators[i] = validators[i];
+    }
+    return miningValidators;
```

Comment:
Try if slice works on solidity

Reply:
It seems that getting slices in solidity does not support the `memory` type, I try to return validator[0:n] will get compiling error